### PR TITLE
TYP: pandas/core/dtypes/dtypes.py

### PR DESCRIPTION
**Pull request by simonjayhawkins**

### CODE REVIEW

File: pandas/core/arrays/integer.py [hunk 1]
```diff
@@ -697,7 +697,7 @@ def integer_arithmetic_method(self, other):
 """
 
 # create the Dtype
-Int8Dtype = register_extension_dtype(
+Int8Dtype: Type[_IntegerDtype] = register_extension_dtype(
```
Comment (simonjayhawkins) [inline pandas/core/arrays/integer.py]: mypy requires type annotations for dynamically created class.

pandas\core\arrays\integer.py:712: error: Need type annotation for 'Int16Dtype'
pandas\core\arrays\integer.py:724: error: Need type annotation for 'Int32Dtype'
pandas\core\arrays\integer.py:736: error: Need type annotation for 'Int64Dtype'
pandas\core\arrays\integer.py:748: error: Need type annotation for 'UInt8Dtype'
pandas\core\arrays\integer.py:760: error: Need type annotation for 'UInt16Dtype'
pandas\core\arrays\integer.py:772: error: Need type annotation for 'UInt32Dtype'
pandas\core\arrays\integer.py:784: error: Need type annotation for 'UInt64Dtype'

a better fix IMO is in #31347, but this will keep mypy green for now.

File: pandas/core/dtypes/dtypes.py [hunk 2]
```diff
@@ -13,10 +25,21 @@
...
+    from pandas.core.arrays import (  # noqa: F401
+        IntervalArray,
+        PeriodArray,
+        DatetimeArray,
+    )
+    from pandas import Categorical  # noqa: F401
+
 str_type = str
 
+ExtensionDtypeT = TypeVar("ExtensionDtypeT", bound=ExtensionDtype)
```
Comment (jorisvandenbossche) [inline pandas/core/dtypes/dtypes.py]: When do you need to use `ExtensionDtypeT` and when can you use `ExtensionDtype` (both are used in this file/diff)

Comment (simonjayhawkins) [reply pandas/core/dtypes/dtypes.py]: not using a typevar for `find` may be an oversight, I'll look into whether necessary.

Comment (simonjayhawkins) [reply pandas/core/dtypes/dtypes.py]: the only callsites are in pandas\core\dtypes\common.py, so `find` may need typevars once types are added there, and the callsites to there are typed etc. etc. 

There does seem to be some resistance to typevars, shall I add now or wait till becomes necessity?

I've also just noticed that the type annotation for `dtype` in `find` is also incorrect...

the code is 
```
            if not isinstance(dtype, type):
                dtype_type = type(dtype)
            if issubclass(dtype_type, ExtensionDtype):
                return dtype
```

so dtype parameter can also accept an instance as well as a type. again could update here or in a subsequent PR. only the return type was changed in this PR as this caused errors once construct_from_string was typed as this returns an instance, return annotation on master is only 
 type.

File: pandas/core/dtypes/dtypes.py [hunk 3]
```diff
@@ -655,39 +678,39 @@ class DatetimeTZDtype(PandasExtensionDtype):
     _match = re.compile(r"(datetime64|M8)\[(?P<unit>.+), (?P<tz>.+)\]")
     _cache: Dict[str_type, PandasExtensionDtype] = {}
 
-    def __init__(self, unit="ns", tz=None):
+    def __init__(self, unit: Union[str_type, "DatetimeTZDtype"] = "ns", tz=None):
```
Comment (WillAyd) [inline pandas/core/dtypes/dtypes.py]: Is this worth an alias in pandas._typing?

File: pandas/core/dtypes/dtypes.py [hunk 4]
```diff
@@ -655,39 +678,39 @@ class DatetimeTZDtype(PandasExtensionDtype):
...
-                tz = result.tz
-                msg = (
-                    f"Passing a dtype alias like 'datetime64[ns, {tz}]' "
-                    "to DatetimeTZDtype is no longer supported. Use "
-                    "'DatetimeTZDtype.construct_from_string()' instead."
-                )
-                raise ValueError(msg)
-            else:
-                raise ValueError("DatetimeTZDtype only supports ns units")
+            self._unit, self._tz = unit.unit, unit.tz
```
Comment (WillAyd) [inline pandas/core/dtypes/dtypes.py]: Just out of curiosity why did this need to change?

**Review by WillAyd — CHANGES_REQUESTED** [review]
question but generally looks good. need to merge master

File: pandas/core/dtypes/dtypes.py (hunk 4 above)
Comment (simonjayhawkins) [reply pandas/core/dtypes/dtypes.py]: This gives `pandas\core\dtypes\dtypes.py:684: error: "str" has no attribute "tz"` which looks weird considering the isinstance check and maybe a mypy issue.

if we split the tuple assignment (which would obviously be wrong), may explain the message

```
        if isinstance(unit, DatetimeTZDtype):
            reveal_type(unit)
            reveal_type(unit.unit)
            reveal_type(unit.tz)
            unit = unit.unit
            tz = unit.tz
            reveal_type(unit)
```

```
pandas\core\dtypes\dtypes.py:683: note: Revealed type is 'pandas.core.dtypes.dtypes.DatetimeTZDtype'
pandas\core\dtypes\dtypes.py:684: note: Revealed type is 'builtins.str'
pandas\core\dtypes\dtypes.py:685: note: Revealed type is 'Any'
pandas\core\dtypes\dtypes.py:687: error: "str" has no attribute "tz"
pandas\core\dtypes\dtypes.py:688: note: Revealed type is 'builtins.str'
```

Comment (WillAyd) [reply pandas/core/dtypes/dtypes.py]: Yea that seems weird. If we just type ignore does anything else need to change? Slight preference for keeping as is and opening an issue on mypy for narrowing issue, if one doesn’t already exist

Comment (simonjayhawkins) [reply pandas/core/dtypes/dtypes.py]: reversing the assignment does not give a false positive, so could do this instead..

```
          tz, unit = unit.tz, unit.unit
```

Comment (simonjayhawkins) [reply pandas/core/dtypes/dtypes.py]: or we could ignore.

Comment (WillAyd) [reply pandas/core/dtypes/dtypes.py]: Reversing works as well but in any case would be helpful to raise with mypy

**Comment by jreback** [comment]
can you rebase

File: pandas/core/dtypes/dtypes.py (hunk 3 above)
Comment (simonjayhawkins) [reply pandas/core/dtypes/dtypes.py]: not sure, is it common to be able to pass a dtype where a string representing a unit is expected?

Comment (jorisvandenbossche) [reply pandas/core/dtypes/dtypes.py]: That is common (almost any place accepting a dtype will also accept a string), but I don't know if there are many places that only accepts a DatetimeTZDtype

File: pandas/core/dtypes/dtypes.py [hunk 5]
```diff
@@ -65,7 +85,7 @@ def register(self, dtype: Type[ExtensionDtype]) -> None:
         """
         Parameters
         ----------
-        dtype : ExtensionDtype
+        dtype : Type[ExtensionDtype]
```
Comment (jorisvandenbossche) [inline pandas/core/dtypes/dtypes.py]: Can you put this in a non-typing way? Eg "ExtensionDtype class"

File: pandas/core/dtypes/dtypes.py [hunk 6]
```diff
@@ -701,20 +721,20 @@ def tz(self):
...
+    def construct_array_type(cls) -> Type["DatetimeArray"]:
         """
         Return the array type associated with this dtype.
 
         Returns
         -------
         type
         """
-        from pandas.core.arrays import DatetimeArray
+        from pandas.core.arrays import DatetimeArray  # noqa: F811
```
Comment (jorisvandenbossche) [inline pandas/core/dtypes/dtypes.py]: Why is this needed now?

File: pandas/core/dtypes/dtypes.py (hunk 3 above)
Comment (simonjayhawkins) [reply pandas/core/dtypes/dtypes.py]: most methods that have a unit parameter, document the unit as string type, so AFAICT, this is the only place where unit can be passed a DatetimeTZDtype instead and the unit extracted.

Comment (simonjayhawkins) [reply pandas/core/dtypes/dtypes.py]: maybe i've misunderstood. I assume the question was about an alias for unit. We already have an alias for Dtype.

File: pandas/core/dtypes/dtypes.py (hunk 6 above)
Comment (simonjayhawkins) [reply pandas/core/dtypes/dtypes.py]: redefinition of unused 'DatetimeArray' from line 29 flake8(F811)

File: pandas/core/dtypes/dtypes.py (hunk 2 above)
Comment (simonjayhawkins) [reply pandas/core/dtypes/dtypes.py]: The typevar has been removed for now.

**Comment by jbrockmendel** [comment]
this looks fine; any sticking points?

File: pandas/core/dtypes/dtypes.py [hunk 7]
```diff
@@ -655,9 +675,9 @@ class DatetimeTZDtype(PandasExtensionDtype):
     _match = re.compile(r"(datetime64|M8)\[(?P<unit>.+), (?P<tz>.+)\]")
     _cache: Dict[str_type, PandasExtensionDtype] = {}
 
-    def __init__(self, unit="ns", tz=None):
+    def __init__(self, unit: Union[str_type, "DatetimeTZDtype"] = "ns", tz=None):
         if isinstance(unit, DatetimeTZDtype):
-            unit, tz = unit.unit, unit.tz
+            unit, tz = unit.unit, unit.tz  # type: ignore
```
Comment (WillAyd) [inline pandas/core/dtypes/dtypes.py]: Can you add comment on why this is needed?

**Review by WillAyd — CHANGES_REQUESTED** [review]

File: pandas/core/dtypes/dtypes.py (hunk 7 above)
Comment (simonjayhawkins) [reply pandas/core/dtypes/dtypes.py]: sure. want to explore that mypy issue further. so maybe a few days

**Comment by WillAyd** [comment]
@simonjayhawkins can you merge master?

**Review by jreback — APPROVED** [review]
yeah merge master and ping on green.

File: pandas/core/dtypes/dtypes.py [hunk 8]
```diff
@@ -13,6 +24,15 @@
 from pandas.core.dtypes.generic import ABCCategoricalIndex, ABCDateOffset, ABCIndexClass
 from pandas.core.dtypes.inference import is_bool, is_list_like
 
+if TYPE_CHECKING:
+    import pyarrow  # noqa: F401
```
Comment (jorisvandenbossche) [inline pandas/core/dtypes/dtypes.py]: This means that running mypy (eg in a pre-commit hook) requires pyarrow to be installed? (which is not a required dependency?) 
Are we fine with that? (do we already do that for other deps?)

Comment (simonjayhawkins) [reply pandas/core/dtypes/dtypes.py]: > This means that running mypy (eg in a pre-commit hook) requires pyarrow to be installed?

no. mypy is a static checker.

The behavior if pyarrow is not installed depends on the strictness of the type checking.(ignore_missing_imports config option)

in general,"mypy will assume the type of that module is Any, the dynamic type. This means attempting to access any attribute of the module will automatically succeed" 
 https://mypy.readthedocs.io/en/stable/running_mypy.html#missing-imports

**Comment by WillAyd** [comment]
I think this was lost in the shuffle @simonjayhawkins . If you can merge master to be safe just merge on green
